Builds/Test.py

```diff
@@ -227,45 +327,168 @@ def run_build(args=None):
+                print('Success')
+
+            if ARGS.clean:
+                shutil.rmtree('build')
+                if '--ninja' in args:
+                    os.remove('build.ninja')
+                    os.remove('.ninja_deps')
+                    os.remove('.ninja_log')
+    else:
+        if ARGS.all:
+            build_dir_targets = CMAKE_DIR_TARGETS
+            generator_options = CMAKE_ALL_GENERATE_OPTIONS
         else:
-            print('Success')
+            build_dir_targets = { tuple(ARGS.dir) : [ARGS.target, ARGS.config] }
+            if ARGS.generator_option:
+                generator_options = [tuple(ARGS.generator_option)]
+            else:
+                generator_options = [tuple()]
 
-        if ARGS.clean:
-            shutil.rmtree('build')
-            if '--ninja' in args:
-                os.remove('build.ninja')
-                os.remove('.ninja_deps')
-                os.remove('.ninja_log')
+        if not build_dir_targets:
+            # Let CMake choose the build tool.
+            build_dir_targets = { () : [] }
+
+        if ARGS.build_option:
+            ARGS.build_option = ARGS.build_option + list(ARGS.extra_args)
+        else:
+            ARGS.build_option = list(ARGS.extra_args)
+
+        for args in generator_options:
+            for build_dirs, (build_targets, build_configs) in build_dir_targets.iteritems():
+                if not build_dirs:
+                    build_dirs = ('default',)
+                if not build_targets:
+                    build_targets = ('rippled',)
+                if not build_configs:
+                    build_configs = ('',)
+                for cmake_dir in build_dirs:
+                    cmake_full_dir = get_cmake_dir(cmake_dir)
+                    run_cmake(cmake_full_dir, cmake_dir, args)
+
+                    for target in build_targets:
+                        for config in build_configs:
+                            run_cmake_build(cmake_full_dir, target, config, ARGS.clean, ARGS.build_option)
+
+                for cmake_dir in build_dirs:
+                    cmake_full_dir = get_cmake_dir(cmake_dir)
+                    for target in build_targets:
+                        for config in build_configs:
+                            failed = run_cmake_tests(cmake_full_dir, target, config)
+
+                            if failed:
+                                print('FAILED:', *(':'.join(f) for f in failed))
+                                if not ARGS.keep_going:
+                                    sys.exit(1)
+                                else:
+                                    all_failed.extend([decodeString(cmake_dir +
+                                            "." + target + "." + config), ':'.join(f)]
+                                        for f in failed)
+                            else:
+                                print('Success')
 
     if all_failed:
-        if len(to_build) > 1:
+        if len(all_failed) > 1:
```

Comment:
Not sure what I was thinking there, probably should have written a comment. I'm not sure if the length check makes sense that way though.